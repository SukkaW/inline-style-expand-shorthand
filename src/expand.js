import expandProperty from './expandProperty'

export default function expand(style) {
  for (const property in style) {
    const value = style[property]

    if (typeof value === 'string' || typeof value === 'number') {
      const expansion = expandProperty(property, value)

      if (expansion) {
        Object.assign(style, expansion)
        delete style[property]
      }
    } else if (value === null) {
      // should skip
    } else if (Array.isArray(value)) {
      if (property === 'extend') {
        value.map(expand)
      } else {
        const expansion = expandProperty(property, value)

        if (expansion) {
          Object.assign(style, expansion)
          delete style[property]
        }
      }
    } else if (typeof value === 'object') {
      expand(value)
    }
  }

  return style
}
