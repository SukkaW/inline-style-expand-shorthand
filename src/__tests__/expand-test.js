import expand from '../expand'

describe('Expanding style objects', () => {
  it('should expand properties in nested objects', () => {
    expect(
      expand({
        width: 200,
        border: '1px solid grey',
        padding: '20px 10px',
        ':hover': {
          borderWidth: '5px 2px 3px 4px',
          margin: '5px',
          height: 100,
        },
      })
    ).toMatchSnapshot()
  })

  it('should not throw on undefined values', () => {
    expect(
      expand({
        border: undefined,
        padding: null,
      })
    ).toMatchSnapshot()
  })

  it('should expand values in arrays', () => {
    expect(
      expand({
        numeralArray: [1, 2],
        padding: [1, '20px'],
        extend: [
          { padding: '10px' },
          { margin: 0 },
          { outline: [0, '10px solid red'] },
        ],
      })
    ).toMatchSnapshot()
  })
})
