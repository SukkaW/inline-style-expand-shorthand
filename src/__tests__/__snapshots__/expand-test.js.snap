// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Expanding style objects should expand properties in nested objects 1`] = `
Object {
  ":hover": Object {
    "borderBottomWidth": "3px",
    "borderLeftWidth": "4px",
    "borderRightWidth": "2px",
    "borderTopWidth": "5px",
    "height": 100,
    "marginBottom": "5px",
    "marginLeft": "5px",
    "marginRight": "5px",
    "marginTop": "5px",
  },
  "borderBottomColor": "grey",
  "borderBottomStyle": "solid",
  "borderBottomWidth": "1px",
  "borderLeftColor": "grey",
  "borderLeftStyle": "solid",
  "borderLeftWidth": "1px",
  "borderRightColor": "grey",
  "borderRightStyle": "solid",
  "borderRightWidth": "1px",
  "borderTopColor": "grey",
  "borderTopStyle": "solid",
  "borderTopWidth": "1px",
  "paddingBottom": "20px",
  "paddingLeft": "10px",
  "paddingRight": "10px",
  "paddingTop": "20px",
  "width": 200,
}
`;

exports[`Expanding style objects should expand values in arrays 1`] = `
Object {
  "extend": Array [
    Object {
      "paddingBottom": "10px",
      "paddingLeft": "10px",
      "paddingRight": "10px",
      "paddingTop": "10px",
    },
    Object {
      "marginBottom": "0",
      "marginLeft": "0",
      "marginRight": "0",
      "marginTop": "0",
    },
    Object {
      "outlineColor": Array [
        "red",
      ],
      "outlineStyle": Array [
        "solid",
      ],
      "outlineWidth": Array [
        "0",
        "10px",
      ],
    },
  ],
  "numeralArray": Array [
    1,
    2,
  ],
  "paddingBottom": Array [
    "1",
    "20px",
  ],
  "paddingLeft": Array [
    "1",
    "20px",
  ],
  "paddingRight": Array [
    "1",
    "20px",
  ],
  "paddingTop": Array [
    "1",
    "20px",
  ],
}
`;

exports[`Expanding style objects should not throw on undefined values 1`] = `
Object {
  "border": undefined,
  "padding": null,
}
`;
